### ISSUE TITLE
Integrate AuditLog to default history page

### ISSUE THREAD

**Issue by wangxuepeng** [jazzband/django-auditlog#68]
> The core idea of Auditlog is similar to the log from Django’s admin. Unlike the log from Django’s admin (django.contrib.admin) Auditlog is much more flexible. Also, Auditlog saves a summary of the changes in JSON format, so changes can be tracked easily.

Considering this design purpose, it would be good to integrate or replace the default history view page with AuditLog log page, this will make the history page info more comprehensive and useful.

The implementation may need create a base ModelAdmin page and let all Admin page inherits it. It's better if the JSON changes can be shown in a beautiful format than the raw data.

Thanks.

**Comment by jjkester** [comment]
Good idea. If anyone wants to tackle this one, please do. (Otherwise it might take some time before I get to it.)

**Comment by Paul424** [comment]
Often i use a simple link to quickly get to related models, something like this could work:

    def history_link(self):
        return u'<a href="{url}?id__exact={id}">{name}</a>'.format(
            url=reverse_lazy("admin:auditlog_logentry_changelist", args=()),
            id=self.id, name='history')

    history_link.allow_tags = True
    history_link.short_description = 'History'

**Comment by shifeitong** [comment]
Hi, I'm new to Auditlog. Why not just extend the `Django admin LogEntry`? Right now, it seems a little redundant to me. Every time I make a change through the admin site, records are generated in both `Auditlog LogEntry` and `Django admin`. A couple of fields are recorded twice.

**Comment by jjkester** [comment]
@shifeitong The Admin site model is very limited and serves a different purpose. It logs only a message, where auditlog will log more details. Also, the the Admin site functionality is very much tied into the admin site. It will only log changes made there.

Keeping both active at the same time gives you the opportunity to see changes that were made by business logic in the application (logged by auditlog) and changes made through the admin (logged by both the admin and auditlog).

Adding a separate "audit" page next to the existing "history" page in the admin would provide the most value.

**Comment by tombreit** [comment]
> @shifeitong The Admin site model is very limited and serves a different purpose. It logs only a message, where auditlog will log more details. Also, the the Admin site functionality is very much tied into the admin site. It will only log changes made there.
> 
> Keeping both active at the same time gives you the opportunity to see changes that were made by business logic in the application (logged by auditlog) and changes made through the admin (logged by both the admin and auditlog).
> 
> Adding a separate "audit" page next to the existing "history" page in the admin would provide the most value.

Good idea, I've extended the admin change form template and added the "Audit" link to Djangos `object_tools` link list:

```html
<!-- path/to/templates/admin/<app_name>/change_form.html -->

{% extends "admin/change_form.html" %}
{% load i18n admin_urls %}

{% block object-tools-items %}
    <li>
        <a href="{% url 'admin:auditlog_logentry_changelist' %}?object_pk={{ original.pk }}">
            {% translate "Audit" %}
        </a>
    </li>
    {{ block.super }}
{% endblock %}
```

**Comment by OscarVanL** [comment]
I got this working with a mix of @joeribekker and @tombreit 's answers. I thought I'd post my solution for completion as it still took a little bit of thinking to get this to work.

1. Create a file `templates/admin/change_form_object_tools.html`. Here we can override the default django "History" button.
Enter something like this:
```
{# Override the default django 'history' button with a custom one based off django-auditlog #}
{# Based off: https://github.com/jazzband/django-auditlog/issues/68 #}

{% load i18n admin_urls %}
{% block object-tools-items %}
    {{ custom_history_button|safe }}
{% endblock %}
```
2. Create a custom `change_view` function in your admin class to add an `extra_content`:
```
class AdminModel(admin.ModelAdmin):
    def change_view(self, request, object_id, form_url='', extra_context=None):
        if extra_context is None:
            extra_context = {}

        # Override the django admin 'History' button to point to the django-auditlog history page.
        row_audit_url = '{url}?object_id={object_id}&resource_type={content_type}'.format(
            url=reverse("admin:auditlog_logentry_changelist", args=()),
            content_type=ContentType.objects.get_for_model(self.model).pk,
            object_id=object_id,
        )

        extra_context['custom_history_button'] = format_html(
            '<a class="button" href="{}">Audit History</a>',
            row_audit_url,
        )
        return super().change_view(request, object_id, form_url, extra_context)
```

Make sure the models you register to admin.site subclass from `AdminModel`.

**Comment by 2ykwang** [comment]
Instead of subclassing/merging into Django’s built-in Admin history, we implemented a separate view and provided a access link. (admin field) 
- (see #743)

This is easier to maintain and avoids unexpected changes to core Admin UX.

Proposing to close — if there are no objections, I’ll close this in 7 days.
If full integration with the built-in admin history is still needed, please open a separate enhancement.


Thanks for the idea @Paul424 @joeribekker @tombreit @OscarVanL
